File: trapz_integral.cpp
#include <iostream>
#include <cmath>
#include <fstream>
#include <chrono>

using namespace std;
using namespace std::chrono;

// Function to integrate
double f(double x) {
    return exp(-x) * cos(100.0 * x) * sqrt(x*x*x + sin(x*x));
}

int main() {
    double a = 0.0;
    double b = 100.0;
    long N = 100000000L;  // points
    double h = (b - a) / (N - 1);
    double sum = 0.0;

    // Start timer
    auto start = high_resolution_clock::now();

    // Trapezoidal integration
    for (long i = 0; i < N; i++) {
        double x = a + h * i;
        if (i == 0 || i == N-1) {
            sum += f(x);
        } else {
            sum += 2.0 * f(x);
        }
    }

    double integral = h * sum / 2.0;

    // Stop timer
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);

    // Output results
    cout.precision(15);
    cout << "Integration result: " << integral << endl;
    cout << "Execution time: " << duration.count() / 1e6 << " seconds" << endl;

    // Export function data for plotting (between 0 and 5)
    ofstream outfile("function_data_cpp.dat");
    for (int i = 0; i < 1000; i++) {
        double x = 0.0 + (5.0 - 0.0) * i / 999.0;
        outfile << x << " " << f(x) << endl;
    }
    outfile.close();

    return 0;
}

File: trapz_integral.f90
program trapz_integral
  implicit none
  integer, parameter :: dp = kind(1.0d0)
  integer, parameter :: long_int = selected_int_kind(18)  ! For large integers
  integer(kind=long_int) :: N, i
  real(dp) :: a, b, h, x, sum, integral
  real(dp) :: t1, t2

  a = 0.0_dp
  b = 100.0_dp
  N = 100000000_long_int  ! points
  h = (b - a) / real(N - 1, dp)
  sum = 0.0_dp

  call cpu_time(t1)

  do i = 1, N
    x = a + h * real(i - 1, dp)
    if (i == 1 .or. i == N) then
      sum = sum + f(x)
    else
      sum = sum + 2.0_dp * f(x)
    end if
  end do

  integral = h * sum / 2.0_dp

  call cpu_time(t2)

  print *, "Result of integral: ", integral
  print *, "Teme of execution: ", t2 - t1, " seconds"

  ! Export data of the function for plot 
  open(unit=10, file="function_data_F90.dat", status="replace")
  do i = 1, 1000
    x = 0.0_dp + (5.0_dp - 0.0_dp) * real(i - 1, dp) / 999.0_dp
    write(10,*) x, f(x)
  end do
  close(10)

contains

  real(dp) function f(x)
    real(dp), intent(in) :: x
    f = exp(-x) * cos(100.0_dp * x) * sqrt(x**3 + sin(x**2))
  end function f

end program trapz_integral
